### DISCUSSION TITLE
How can I define a custom comparator when ordering fields eventually coming from a custom Java type ?

### DISCUSSION THREAD

**Discussion by facewindu** [jOOQ/jOOQ#17154] [Q&A] [answered]
I have a custom java type, which is called `VersionNumber`. It implements the `Comparable` interface, and also has nice `fromString`/`toString` methods.

Following string logic, "8.9" > "8.10", but really, when those are versions "8.9" < "8.10" (which my `VersionNumber` type knows about and therefore has the right comparison logic).

Now, in jOOQ, we currently store those in fields by using the string representation of `VersionNumber`.
Naturally, when we `orderBy` this field, results come out sorted by "string-logic".

Is there a simple way to instruct jOOQ that it should order them by converting them to `VersionNumber` and use the `VersionNumber#compareTo` method instead ?

I tried (pseudo code) to have my `SortField` created as
```
myTable.myVersionNumberStringField.convertFrom(VersionNumber::parse) // This returns a Field<VersionNumber>
```
but this is not enough.

How can I tell jOOQ how to compare `Field<VersionNumber>` fields ?

I saw [this](https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-order-provider/) but it seemed too high-level, or I didn't really understand how to apply it on fields.

**Comment by lukaseder** [comment, marked as answer]
Thanks for your message. You cannot use Java code to order things in SQL. Your SQL database wouldn't know how to access your `compareTo()` method. I know with jOOQ, these things look like they're the same, but it's important to distinguish between Java code that is executed in Java and Java/jOOQ code that is just an expression of logic that will be executed in SQL.

You didn't mention your SQL dialect. If this is in PostgreSQL, you could do this:

```sql
select v, string_to_array(v, '.')::int[] a
from (values ('0.9'), ('0.10'), ('0.10.1')) as t (v)
order by a
```

Of course, you could sort results *after* fetching them from SQL (e.g. using `Result.sort()` methods), but that has a very different semantics. E.g. it doesn't interact with the SQL `LIMIT` clause, for example.

I hope this helps.

**Reply by facewindu** [reply]
yes it does, thanks
